fem/gridfunc.cpp

```diff
@@ -4634,4 +4634,801 @@ GridFunction *Extrude1DGridFunction(Mesh *mesh, Mesh *mesh2d,
    return sol2d;
 }
 
+void GridFunction::SetupPLBounds(const int nb, const int ncp,
+                                 const int b_type, const int cp_type,
+                                 const real_t tol,
+                                 Vector &nodex, Vector &nodew, Vector &intx,
+                                 DenseMatrix &lbound, DenseMatrix &ubound)
+{
+   lbound.SetSize(nb, ncp);
+   ubound.SetSize(nb, ncp);
+   nodex.SetSize(nb);
+   nodew.SetSize(nb);
+   intx.SetSize(ncp);
+   auto scalenodes = [](const Vector &in, const real_t a, const real_t b) -> Vector
+   {
+      Vector outVec(in.Size());
+      real_t maxv = in.Max();
+      real_t minv = in.Min();
+      for (int i = 0; i < in.Size(); i++)
+      {
+         outVec(i) = a + (b-a)*(in(i)-minv)/(maxv-minv);
+      }
+      return outVec;
+   };
+   MFEM_VERIFY(ncp >= 2,"Atleast 2 control points are required.");
+
+   if (cp_type == 0) // GL + End Point
+   {
+      intx(0) = 0.0;
+      intx(ncp-1) = 1.0;
+      if (ncp > 2)
+      {
+         const real_t *x = poly1d.GetPoints(ncp-3, 0);
+         MFEM_VERIFY(x, "Error in getting points.");
+         for (int i = 0; i < ncp-2; i++)
+         {
+            intx(i+1) = x[i];
+         }
+      }
+   }
+   else if (cp_type == 1) // Chebyshev
+   {
+      auto GetChebyshevNodes = [](int n) -> Vector
+      {
+         Vector nodes(n);
+         nodes(0) = -1.0;
+         nodes(n - 1) = 1.0;
+         for (int i = 2; i < n; ++i)
+         {
+            nodes(i - 1) = -std::cos(M_PI * ((i - 1.0)*1.0 / (n - 1)));
+         }
+         return nodes;
+      };
+      intx = GetChebyshevNodes(ncp);
+   }
+   else
+   {
+      MFEM_ABORT("Unsupported interval points. Use [0,1].\n");
+   }
+   intx = scalenodes(intx, 0.0, 1.0); // rescale to [0,1]
+
+   MFEM_VERIFY(b_type >= 0 && b_type <= 2, "Invalid node type, use [0,2].");
+   Poly_1D::Basis &basis1d(poly1d.GetBasis(nb-1, b_type));
+
+   // Initialize bounds
+   lbound = 0.0;
+   ubound = 0.0;
+
+   Vector bmv(nb), bpv(nb), bv(nb); // basis values
+   Vector bdmv(nb), bdpv(nb), bdv(nb); // basis derivative values
+   Vector vals(3);
+
+   for (int j = 0; j < ncp; j++)
```

Comment:
A comment or two here might be nice to explain what this corresponds to in the papers